Support Cross NS Spark Write and Fix Cross NS authorization tests

rohitsinha54 opened the pull request:
- Issue: https://issues.cask.co/browse/CDAP-6793 and https://issues.cask.co/browse/CDAP-6339
- Build: http://builds.cask.co/browse/CDAP-RUT48-1
## Summary
- Adds support for cross namespace write in Spark
- Refactored and cleaned up unit tests written for cross namespace read for spark here https://github.com/caskdata/cdap/pull/6036 and remove copied apps written for cross namespace access.
- All Unit tests written for cross namespace authorization tests for flow, mr and spark were incorrect (PR: https://github.com/caskdata/cdap/pull/6261). They were running program as a user who does not have any right in the namespace where the app was deployed rather than allowing user starting the app to have access to it and no access to  cross namespace input/output. Hence, all the assertions were wrong too as earlier will cause start on program to throw exception but in actual cross namespace access will lead to runtime failure after the program is running.

Commit:
CDAP-6793 Support cross namespace write in Spark

Commit:
CDAP-6339  fix AuthorizationTest for cross namespace access

bdmogal commented on cdap-api-spark/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContext.java:
```diff
@@ -445,6 +445,18 @@
+   *
+   * @param rdd the {@link JavaPairRDD} to be saved
+   * @param namespace the namespace for the dataset
+   * @param datasetName name of the Dataset
```
In this file, I think you can remove capitalization of 'D'ataset when its not an `{@link }` element

bdmogal commented on cdap-api-spark/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContext.java:
```diff
@@ -445,6 +445,18 @@
+   * Saves the given {@link JavaPairRDD} to the given {@link Dataset}.
+   *
+   * @param rdd the {@link JavaPairRDD} to be saved
+   * @param namespace the namespace for the dataset
```
the namespace in which the specified dataset is to be saved

bdmogal commented on cdap-api-spark/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContext.java:
```diff
@@ -453,4 +465,16 @@
+   * Saves the given {@link JavaPairRDD} to the given {@link Dataset} with the given set of Dataset arguments.
+   *
+   * @param rdd the {@link JavaPairRDD} to be saved
+   * @param namespace the namespace for the Dataset
```
the namespace in which the specified dataset is to be saved

bdmogal commented on cdap-api-spark/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContext.java:
```diff
@@ -453,4 +465,16 @@
+   * @param namespace the namespace for the Dataset
+   * @param datasetName name of the Dataset
+   * @param arguments arguments for the Dataset
+   * @throws DatasetInstantiationException if the Dataset doesn't exist
```
what are the semantics of this method? If the dataset doesn't exist, you should be able to create a new one, no?

rohitsinha54 replied:
I don't think we support creating new dataset on demand from spark program. The dataset has to exists either created by app configure() during deployment or by user. Here I have just moved the code around so maybe @chtyim can give more information about the semantics.

bdmogal commented on cdap-api-spark/src/main/scala/co/cask/cdap/api/spark/SparkExecutionContext.scala:
```diff
@@ -246,4 +244,17 @@ trait SparkExecutionContext extends RuntimeContext with Transactional {
+    * Saves the given [[org.apache.spark.rdd.RDD]] to the given [[co.cask.cdap.api.dataset.Dataset]].
+    * Using the implicit object [[co.cask.cdap.api.spark.SparkMain.SparkProgramRDDFunctions]] is preferred.
+    *
+    * @param rdd         the [[org.apache.spark.rdd.RDD]] to operate on
```
extra spaces can be removed for all parameters in this javadoc

bdmogal commented on cdap-api-spark/src/main/scala/co/cask/cdap/api/spark/SparkMain.scala:
```diff
@@ -112,14 +112,53 @@ trait SparkMain extends Serializable {
+      * @param namespace the namespace for the dataset
+      * @param datasetName name of the Dataset
+      * @param sec the [[co.cask.cdap.api.spark.SparkExecutionContext]] of the current execution
+      * @throws co.cask.cdap.api.data.DatasetInstantiationException if the Dataset doesn't exist
```
same question about semantics here.

bdmogal commented on cdap-api-spark/src/main/scala/co/cask/cdap/api/spark/SparkMain.scala:
```diff
@@ -112,14 +112,53 @@ trait SparkMain extends Serializable {
+      * Saves the given [[org.apache.spark.rdd.RDD]] to the given [[co.cask.cdap.api.dataset.Dataset]].
+      *
+      * @param namespace the namespace for the dataset
+      * @param datasetName name of the Dataset
```
nit: reduce capitalization

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -876,36 +881,64 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
+    // switch to BOB
+    SecurityRequestContext.setUserId(BOB.getName());
+
+    // Send data to stream as BOB this ensures that BOB can do stuff in auth namespace
```
nit: can write to a stream in the auth namespace?

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -876,36 +881,64 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
     DataSetManager<KeyValueTable> dataSetManager = getDataset(datasetOutputSpace.toId(), "store");
     KeyValueTable results = dataSetManager.get();
+
+    CloseableIterator<KeyValue<byte[], byte[]>> scan = results.scan(null, null);
```
have to close this

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -931,72 +964,122 @@ public void testCrossNSDatasetAccessWithAuthMapReduce() throws Exception {
-    } catch (Exception e) {
-      // Expected
-    }
+    verifyProgramFailure(argsForMR, mrManager);
```
you can reuse these helper methods in the previous test as well right?

rohitsinha54 replied:
In Flowlets, we are testing for its status to go into running and then call stop on it which will make the status of run killed, since its a realtime program. Unlike mapreduce and spark where the program stops when the runtime exception happens and the status is marked as failed.

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -931,72 +964,122 @@ public void testCrossNSDatasetAccessWithAuthMapReduce() throws Exception {
-    } catch (Exception e) {
-      // Expected
-    }
+    ImmutableMap<String, String> args = ImmutableMap.of(
```
just Map

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkTestRun.java:
```diff
@@ -309,21 +308,36 @@ public void testScalaSparkWithObjectStore() throws Exception {
+    // deploy the spark app in another namespace (default)
     ApplicationManager applicationManager = deploy(SparkAppUsingObjectStore.class);
+
+    ImmutableMap<String, String> args = ImmutableMap.of(ScalaCrossNSProgram.STREAM_NAMESPACE(),
```
just Map

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/spark/SparkTestRun.java:
```diff
@@ -334,16 +348,19 @@ public void testScalaSparkCrossNSStream() throws Exception {
+    DataSetManager<ObjectStore<String>> keysManager = getDataset(inputDSNSMeta.getNamespaceId().toId(), "keys");
     prepareInputData(keysManager);
 
+    ImmutableMap<String, String> args = ImmutableMap.of(ScalaCharCountProgram.INPUT_DATASET_NAMESPACE(),
```
just Map

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/spark/stream/SparkStreamIntegrationTestRun.java:
```diff
@@ -57,38 +57,58 @@ public void testSparkWithStream() throws Exception {
+    // deploy TestSparkStreamIntegrationApp in default namespace
+    ApplicationManager spark1 = deployApplication(TestSparkStreamIntegrationApp.class);
+
+    ImmutableMap<String, String> args = ImmutableMap.of(
```
Map

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/spark/stream/TestSparkCrossNSDatasetApp.java:
```diff
@@ -17,48 +17,60 @@
     @Override
     public void configure() {
-      setName("SparkStreamProgram");
+      setName("SparkCrossNSDatasetProgram");
```
Can use `SparkCrossNSDatasetProgram.class.getSimpleName()`?

bdmogal commented on cdap-unit-test/src/test/scala/co/cask/cdap/spark/app/ScalaCharCountProgram.scala:
```diff
@@ -21,8 +21,12 @@ import co.cask.cdap.api.spark.{AbstractSpark, SparkExecutionContext, SparkMain}
 /**
-  * A simple Spark program written in Scala which counts the number of characters
+  * A simple Spark program written in Scala which counts the number of characters. It reads from a dataset and writes
+  * to another dataset. These datasets can be another namespace and their names should be specified as runtime
```
can be _in_ another namesapce

rohitsinha54 commented:
@bdmogal : Addressed comments. Also since a lot of code has changed in AuthorizationTest in this PR: https://github.com/caskdata/cdap/pull/6427 to avoid merge conflicts I have addressed your comments for any existing code in that PR in separate comment for easier look here: https://github.com/caskdata/cdap/pull/6427/commits/3b92c2fe832d4faf5bd377a550222e066c6cf6c7

rohitsinha54 commented:
New Build: http://builds.cask.co/browse/CDAP-RUT48-2
ITM: http://builds.cask.co/browse/CDAP-ITM0-9

bdmogal commented on cdap-api-spark/src/main/scala/co/cask/cdap/api/spark/SparkMain.scala:
```diff
@@ -111,15 +111,54 @@ trait SparkMain extends Serializable {
+      * @throws co.cask.cdap.api.data.DatasetInstantiationException if the dataset doesn't exist
       */
-    def saveAsDataset(datasetName: String, arguments: Map[String, String] = Map())
+    def saveAsDataset(datasetName: String, arguments: Map[String, String])
```
why remove the `= Map()`? Does this change the behavior for this API in any way?

rohitsinha54 replied:
For this reason: https://github.com/caskdata/cdap/pull/6036#discussion_r71249342

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -1058,4 +1141,22 @@ public boolean apply(Privilege input) {
+  private void verifyDatasetIsEmpty(NamespaceId namespaceId, String datasetName) throws Exception {
+    DataSetManager<KeyValueTable> outTableManager = getDataset(namespaceId.toId(), datasetName);
+    KeyValueTable outputTable = outTableManager.get();
+    Assert.assertFalse(outputTable.scan(null, null).hasNext());
```
scan should be closed.

rohitsinha54 replied:
Addressed here: https://github.com/caskdata/cdap/pull/6427/files#diff-b851802e71aeb1eaaa4e26252c515101R1314
Unless there is strong objection against the change coming in another PR which will follow this one I prefer not to do rework here.

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -876,36 +881,64 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
+    Assert.assertFalse(scan.hasNext());
+
+    // Give BOB permission to write to the dataset in another namespace
+    grantAndAssertSuccess(new DatasetId(datasetOutputSpace.getNamespace(), "store"),
```
just datasetOutputSpace.dataset("store")

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -931,72 +964,122 @@ public void testCrossNSDatasetAccessWithAuthMapReduce() throws Exception {
 
     ApplicationManager spark = deployApplication(AUTH_NAMESPACE.toId(), TestSparkCrossNSDatasetApp.class);
-    SparkManager sparkManager = spark.getSparkManager("SparkStreamProgram");
+    SparkManager sparkManager = spark.getSparkManager("SparkCrossNSDatasetProgram");
```
can this be a reference to some class?

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -931,72 +964,122 @@ public void testCrossNSDatasetAccessWithAuthMapReduce() throws Exception {
+    verifyDatasetIsEmpty(outputDatasetNSMeta.getNamespaceId(), "output");
+
+    // give privilege to BOB on the input dataset
+    grantAndAssertSuccess(new DatasetId(inputDatasetNSMeta.getNamespaceId().getNamespace(), "input"), BOB,
```
inputDatasetNSMeta.getNamespaceId().dataset("input")

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -931,72 +964,122 @@ public void testCrossNSDatasetAccessWithAuthMapReduce() throws Exception {
-    // Run again with Alice who has the privileges
-    sparkManager.start();
+    // give privilege to BOB on the output dataset
+    grantAndAssertSuccess(new DatasetId(outputDatasetNSMeta.getNamespaceId().getNamespace(), "output"), BOB,
```
ditto

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -1058,4 +1141,22 @@ public boolean apply(Privilege input) {
+  private void verifyProgramFailure(Map<String, String> programArgs, ProgramManager programManager)
+    throws TimeoutException, InterruptedException {
+    programManager.start(programArgs);
+    programManager.waitForFinish(5, TimeUnit.MINUTES);
```
I think there is a race between this guy and the next verification on line 1159. When this succeeds, and you move on, the run record may not have been stamped yet. Should wait for it.

rohitsinha54 replied:
Nice catch. Addressed this comment and all other comments on AuthorizationTest class in this commit: https://github.com/caskdata/cdap/pull/6427/commits/e89eec2b2ce5137d41d4f9215cfd45c1723a0036

bdmogal replied:
On that note, can you report a jira to add this run record check as part of `waitForFinish`

chtyim commented on cdap-api-spark/src/main/scala/co/cask/cdap/api/spark/SparkMain.scala:
```diff
@@ -162,12 +201,12 @@ trait SparkMain extends Serializable {
       *
-      * @param datasetName name of the Dataset
-      * @param arguments arguments for the Dataset; default is an empty [[scala.collection.Map]]
+      * @param datasetName name of the dataset
```
Why changing all the cases? I thought we also use `Dataset` to refer to the CDAP dataset?

rohitsinha54 replied:
@bdmogal ^

bdmogal replied:
@chtyim I thought there was a documentation task to reduce this upper-casing. I think if we just upper-case it without an `{@link Dataset}`, it looks a little strange.

rohitsinha54 commented:
Reverted the changes from Dataset to dataset.
